### PR TITLE
Cleaning up the setup instructions for Kubernetes 1.7

### LINKED ISSUE #1004: The Helm Install docs in the API Aggregation Setup don't work on a Mac
The `helm install` command in [this section](https://github.com/kubernetes-incubator/service-catalog/blob/master/docs/api-aggregation-setup.md#install-the-service-catalog-chart-with-helm) don't work on a mac. They produce the following error output:

```console
admins-MacBook-Pro:service-catalog aaschles$ helm install charts/catalog     --name ${HELM_NAME} --namespace ${SVCCAT_NAMESPACE}     --set apiserver.auth.enabled=true         --set useAggregator=true         --set apiserver.tls.ca=$(base64 --wrap 0 ${SC_SERVING_CA})         --set apiserver.tls.cert=$(base64 --wrap 0 ${SC_SERVING_CERT})         --set apiserver.tls.key=$(base64 --wrap 0 ${SC_SERVING_KEY})
base64: unrecognized option `--wrap'
Usage:	base64 [-hvD] [-b num] [-i in_file] [-o out_file]
  -h, --help     display this message
  -D, --decode   decodes input
  -b, --break    break encoded string into num character lines
  -i, --input    input file (default: "-" for stdin)
  -o, --output   output file (default: "-" for stdout)
base64: unrecognized option `--wrap'
Usage:	base64 [-hvD] [-b num] [-i in_file] [-o out_file]
  -h, --help     display this message
  -D, --decode   decodes input
  -b, --break    break encoded string into num character lines
  -i, --input    input file (default: "-" for stdin)
  -o, --output   output file (default: "-" for stdout)
base64: unrecognized option `--wrap'
Usage:	base64 [-hvD] [-b num] [-i in_file] [-o out_file]
  -h, --help     display this message
  -D, --decode   decodes input
  -b, --break    break encoded string into num character lines
  -i, --input    input file (default: "-" for stdin)
  -o, --output   output file (default: "-" for stdout)
Error: a release named "catalog" already exists.
Please run: helm ls --all "catalog"; helm del --help
```

### PR DESCRIPTION
**Pull request by arschles**
This is a multi-part PR:

- Moves the aggregated API TLS setup script into the charts/catalog directory
- Fetches TLS file contents directly from the helm chart to remove the dependency on base64
    - Fixes #1004
- Adds `--authentication-skip-lookup=true` to work around the lack of the `requestheader-client-ca` key in the `extension-apiserver-authentication` ConfigMap in the `kube-system` namespace
- Adds a walkthrough doc for Kube 1.7

### CODE REVIEW

File: charts/catalog/aggregation-tls-setup.sh [hunk 1]
```diff
@@ -14,6 +14,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+CUR=$(pwd)
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
Comment (MHBauer) [inline charts/catalog/aggregation-tls-setup.sh]: at the top change shebang to bash.

File: .gitignore [hunk 2]
```diff
@@ -20,3 +20,11 @@ contrib/build/*/tmp/*
 .pkg
 .kube
 .var
+charts/catalog/apiserver-key.pem
+charts/catalog/apiserver.csr
+charts/catalog/apiserver.pem
```
Comment (MHBauer) [inline .gitignore]: can we put these in a subdir to isolate them?

File: charts/catalog/templates/apiserver-cert-secret.yaml [hunk 3]
```diff
@@ -9,8 +9,8 @@ metadata:
     heritage: "{{ .Release.Service }}"
 type: Opaque
 data:
-  tls.crt: {{ .Values.apiserver.tls.cert }}
-  tls.key: {{ .Values.apiserver.tls.key }}
+  tls.crt: {{ (.Files.Get .Values.apiserver.tls.certFileName) | b64enc }}
+  tls.key: {{ (.Files.Get .Values.apiserver.tls.keyFileName) | b64enc }}
```
Comment (MHBauer) [inline charts/catalog/templates/apiserver-cert-secret.yaml]: is `b64enc` a helm thing?

File: charts/catalog/templates/apiregistration.yaml [hunk 4]
```diff
@@ -13,6 +13,10 @@ spec:
   service:
     namespace: {{ .Release.Namespace }}
     name: {{ template "fullname" . }}-apiserver
-  caBundle: {{ .Values.apiserver.tls.ca }}
+  caBundle: {{ (.Files.Get .Values.apiserver.tls.caFileName) | b64enc }}
   priority: 100
```
Comment (MHBauer) [inline charts/catalog/templates/apiregistration.yaml]: I think this needs to be in the else of the below stuff

File: charts/catalog/values.yaml [hunk 5]
```diff
@@ -9,13 +9,22 @@ apiserver:
...
+    keyFileName:
+    # The name of the file that contains the CA to authenticate connections from API server
+    # proxies. This field is not required. If it is not set, the service-catalog API server 
+    # will be started with the --authentication-skip-lookup flag, which will disable the auth checks
```
Comment (MHBauer) [inline charts/catalog/values.yaml]: I'm not sure setting CA should tie to setting this other flag.

File: docs/walkthrough-1.7.md [hunk 6]
```diff
@@ -0,0 +1,489 @@
+# Service Catalog Demonstration Walkthrough for Kubernetes v1.7.0 and Above
```
Comment (MHBauer) [inline docs/walkthrough-1.7.md]: I know we wanted a separate walkthrough, but I figured it would be for the older versions. Is it expected to look at this one or the existing one when first starting?

File: docs/api-aggregation-setup.md [hunk 7]
```diff
@@ -211,5 +212,12 @@ Before installing the helm chart, run the script in contrib by
...
 ```shell
-source /contrib/svc-cat-apiserver-aggregation-tls-setup.sh
+source charts/catalog/aggregation/tls-setup/sh
+```
+
+To delete, all resources, execute the following:
+
+```console
+helm delete --purge catalog
+k delete apiservice v1alpha1.servicecatalog.k8s.io
```
Comment (MHBauer) [inline docs/api-aggregation-setup.md]: not sure we can expect everyone to have kubectl aliased to k. Can we explicitly right out kubectl?

File: charts/catalog/templates/apiserver-cert-secret.yaml (hunk 3 above)
Comment (arschles) [inline charts/catalog/templates/apiserver-cert-secret.yaml]: yes, it uses [sprig](https://github.com/masterminds/sprig) for template functions. relevant docs: https://masterminds.github.io/sprig/encoding.html

File: .gitignore (hunk 2 above)
Comment (arschles) [inline .gitignore]: I'm not sure, I know this works at the top level chart directory though, and I'd rather make progress this way for now

File: charts/catalog/templates/apiregistration.yaml (hunk 4 above)
Comment (arschles) [inline charts/catalog/templates/apiregistration.yaml]: why?

File: charts/catalog/values.yaml (hunk 5 above)
Comment (arschles) [inline charts/catalog/values.yaml]: they seem to be one or the other

File: charts/catalog/templates/apiregistration.yaml (hunk 4 above)
Comment (MHBauer) [inline charts/catalog/templates/apiregistration.yaml]: It doesn't seem to exist on the object in the beta version.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1/types.go#L39-L72

There's a note too
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1/types.go#L71

File: docs/walkthrough-1.7.md (hunk 6 above)
Comment (arschles) [inline docs/walkthrough-1.7.md]: That would depend on the target Kubernetes cluster. We need to do a better job in directing users to the right doc in the README, but I'd like to do that in a follow-up

File: docs/walkthrough-1.7.md [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+
+## Step 0 - Prerequisites
+
+### Starting Kubernetes with DNS
+
+You *must* have a Kubernetes cluster with cluster DNS enabled. We can't list
+instructions here for enabling cluster DNS for all Kubernetes cluster
+installations, but here are a few notes:
+
+* If you are using Google Container Engine or minikube, you likely have cluster
```
Comment (luxas) [inline docs/walkthrough-1.7.md]: or kubeadm

File: docs/walkthrough-1.7.md [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+[README.md](https://github.com/kubernetes-incubator/service-catalog/blob/master/charts/catalog/README.md).
+
+### The Service Catalog Database
+
+We'll be interacting with a variety of resources in the following steps. The service catalog API
+server needs to store all of these resources in a database. The database implementation in the API
+server is pluggable, and we currently support the following implementations:
+
+1. Etcd 3
+2. Third Party Resources (also, known as TPRs) - this is an _alpha_ feature right now. It has known issues
```
Comment (luxas) [inline docs/walkthrough-1.7.md]: will be migrated to CRD + link to issue

File: docs/walkthrough-1.7.md [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+achieved through the `kubectl` command line interface. Chances are high that
+you already have this installed, however, the service catalog *requires*
+`kubectl` version 1.7 or newer.
+
+To proceed, we must download and install `kubectl` version 1.7 or newer. Do so by following the 
+[standard instructions](https://kubernetes.io/docs/tasks/kubectl/install/).
+
+For example, on a mac,
+```console
+curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.7.0/bin/darwin/amd64/kubectl
```
Comment (luxas) [inline docs/walkthrough-1.7.md]: `dl.k8s.io` points to `storage.googleapis.com/kubernetes-release`, can use that instead...

File: docs/walkthrough-1.7.md [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+No resources found
+```
+
+Note that you _must_ reference `Binding`s by their fully-qualified name: 
+`binding.servicecatalog.k8s.io`.
+
+Create the new `Broker` resource with the following command:
+
+```console
+kubectl create -f contrib/examples/walkthrough/ups-broker.yaml
```
Comment (luxas) [inline docs/walkthrough-1.7.md]: Soon this will be in the helm chart: #980

File: charts/catalog/templates/apiserver-deployment.yaml [hunk 12]
```diff
@@ -59,6 +59,8 @@ spec:
         - "{{ .Values.apiserver.verbosity }}"
         {{- if .Values.apiserver.tls.requestHeaderCA }}
         - --requestheader-client-ca-file=/var/run/kubernetes-service-catalog/requestheader-ca.crt
+        {{ else }}
+        - --authentication-skip-lookup=true
```
Comment (luxas) [inline charts/catalog/templates/apiserver-deployment.yaml]: Hmm, I think it's possible to use delegated authn/authz without passing the requestheader-ca when the correct RBAC bindings are in place. See #936 and #981

File: charts/catalog/values.yaml [hunk 13]
```diff
@@ -9,13 +9,22 @@ apiserver:
...
+    # The name of the file that contains the certificate authority for the APIService object to 
+    # register the Service Catalog API with the API aggregator in the APIRegistration object.
+    # This field is required if apiserver.auth.enabled is true.
+    caFileName:
+    # The name of the file that contains the x509 certificate for use in the aforementioned
+    # APIService object. This field is required if apiserver.auth.enabled is true.
```
Comment (luxas) [inline charts/catalog/values.yaml]: I don't think so. RBAC and delegated authn/authz can happen while still serving with self-signed certs

File: charts/catalog/values.yaml (hunk 5 above)
Comment (luxas) [inline charts/catalog/values.yaml]: Agree with @MHBauer

File: docs/api-aggregation-setup.md [hunk 14]
```diff
@@ -179,12 +178,14 @@ keys we just generated inline.
...
 helm install charts/catalog \
-    --name ${HELM_NAME} --namespace ${SVCCAT_NAMESPACE} \
+    --name ${HELM_NAME} \
+    --namespace ${SVCCAT_NAMESPACE} \
     --set apiserver.auth.enabled=true \
-        --set useAggregator=true \
-        --set apiserver.tls.ca=$(base64 --wrap 0 ${SC_SERVING_CA}) \
-        --set apiserver.tls.cert=$(base64 --wrap 0 ${SC_SERVING_CERT}) \
-        --set apiserver.tls.key=$(base64 --wrap 0 ${SC_SERVING_KEY})
+    --set useAggregator=true \
```
Comment (luxas) [inline docs/api-aggregation-setup.md]: I think we should just switch this to true by default, see #981

File: docs/api-aggregation-setup.md [hunk 15]
```diff
@@ -179,12 +178,14 @@ keys we just generated inline.
 
 ```
 helm install charts/catalog \
-    --name ${HELM_NAME} --namespace ${SVCCAT_NAMESPACE} \
+    --name ${HELM_NAME} \
+    --namespace ${SVCCAT_NAMESPACE} \
     --set apiserver.auth.enabled=true \
```
Comment (luxas) [inline docs/api-aggregation-setup.md]: I think we should just switch this to true by default, see #981

File: docs/api-aggregation-setup.md [hunk 16]
```diff
@@ -179,12 +178,14 @@ keys we just generated inline.
...
-        --set useAggregator=true \
-        --set apiserver.tls.ca=$(base64 --wrap 0 ${SC_SERVING_CA}) \
-        --set apiserver.tls.cert=$(base64 --wrap 0 ${SC_SERVING_CERT}) \
-        --set apiserver.tls.key=$(base64 --wrap 0 ${SC_SERVING_KEY})
+    --set useAggregator=true \
+    --set apiserver.insecure=false \
```
Comment (luxas) [inline docs/api-aggregation-setup.md]: I think we should just switch this to false by default, see #981

File: charts/catalog/templates/apiregistration.yaml (hunk 4 above)
Comment (arschles) [inline charts/catalog/templates/apiregistration.yaml]: @MHBauer seems like it would need to be under an if for `v1alpha1`.

Comment (MHBauer) [inline charts/catalog/templates/apiregistration.yaml]: that's fine too.

File: charts/catalog/templates/apiregistration.yaml [hunk 17]
```diff
@@ -13,6 +13,11 @@ spec:
...
     namespace: {{ .Release.Namespace }}
     name: {{ template "fullname" . }}-apiserver
-  caBundle: {{ .Values.apiserver.tls.ca }}
+  caBundle: {{ (.Files.Get .Values.apiserver.tls.caFileName) | b64enc }}
+  {{ if .Capabilities.APIVersions.Has "apiregistration.k8s.io/v1alpha1" -}}
   priority: 100
+  {{ else if .Capabilities.APIVersions.Has "apiregistration.k8s.io/v1beta1" -}}
+  groupPriorityMinimum: 10000
+  versionPriority: 20
+  {{- end }}
```
Comment (MHBauer) [inline charts/catalog/templates/apiregistration.yaml]: @arschles can we pull this out into a separate PR (without the filesget -> base64 change) ? I think it's worth having while we figure out the details on everything else.

Comment (arschles) [inline charts/catalog/templates/apiregistration.yaml]: @MHBauer I will not be able to split this out today. Please feel free to pull this into another PR, and I'll rebase as necessary.

Comment (MHBauer) [inline charts/catalog/templates/apiregistration.yaml]: https://github.com/kubernetes-incubator/service-catalog/pull/1011

Comment (arschles) [inline charts/catalog/templates/apiregistration.yaml]: thanks @MHBauer

**Comment by arschles** [comment]
@luxas @MHBauer There are tweaks requested and also a request to wait until #936 and #981 are in. I'd like to get this PR in now, and address the tweaks and follow-ups in those PRs.

**Comment by luxas** [comment]
I'd like the tweaks/comments to be adressed and see the CI passing, but other than that it's fine by me to merge if follow-ups to enhance this are promised

File: docs/api-aggregation-setup.md (hunk 14 above)
Comment (pmorie) [inline docs/api-aggregation-setup.md]: I think it would be fine for #981 to do that.

File: docs/walkthrough-1.7.md (hunk 6 above)
Comment (pmorie) [inline docs/walkthrough-1.7.md]: I'm fine with sorting out directing users in a follow-up.

**Comment by pmorie** [comment]
@arschles jenkins looks like a legit error:

<details>
```
1. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command failed. Will retry in 5 seconds.
2. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command failed. Will retry in 7 seconds.
3. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command failed. Will retry in 11 seconds.
4. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command failed. Will retry in 16 seconds.
5. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command failed. Will retry in 25 seconds.
6. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command failed. Will retry in 37 seconds.
7. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command failed. Will retry in 56 seconds.
8. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command failed. Will retry in 60 seconds.
9. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command failed. Will retry in 60 seconds.
10. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer
Error: render error in "catalog/templates/apiserver-cert-secret.yaml": template: catalog/templates/apiserver-cert-secret.yaml:12:33: executing "catalog/templates/apiserver-cert-secret.yaml" at <.Values.apiserver.tl...>: wrong type for value; expected string; got interface {}
Command 'helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/catalog --name catalog --namespace catalog --set debug=true,insecure=true,controllerManager.image=gcr.io/service-catalog-jenkins-jobs/catalog/controller-manager:01cdcb1,apiserver.image=gcr.io/service-catalog-jenkins-jobs/catalog/apiserver:01cdcb1,apiserver.service.type=LoadBalancer' failed 10 times, aborting.
/var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/contrib/hack/test_walkthrough.sh: line 147: Error deploying service catalog to cluster. (exit 1)
```
</details>

**Comment by arschles** [comment]
Closing in favor of #1163
